File: src/seat.lua
-- imports
local posix = require("posix")

-- local imports
local xkbcommon = require("..include.xkbcommon")
local table_helpers = require("table_helpers")
local globals = require("globals")
local config = require("config")
local wau = require("wau")
wau:require("..include.protocol.river-layer-shell-v1")

local m = {}
-- define the seat
m.Seat = { mt = {}, listener = {} }
m.Seat.mt.__index = m.Seat

-- create the seat
function m.Seat.create(obj)
	local seat = {
		obj = obj,
		new = true,
		xkb_bindings = {},
		pointer_bindings = {},
	}
	setmetatable(seat, m.Seat.mt)
	obj:set_user_data(seat)
	obj:add_listener(m.Seat.listener)
	return seat
end

-- seat focus request
function m.Seat:focus(window_local)
    local wm = require("wm")
	if window_local == nil and #wm.wm.windows > 0 then
		-- Fall back to topmost window
	    window_local = wm.wm.windows[#wm.wm.windows]
	end

	if window_local then
		if self.focused ~= window_local then
			self.obj:focus_window(window_local.obj)
			self.focused = window_local
			-- Move to top
			local i = table_helpers.table_index_of(wm.wm.windows, window_local)
			table.remove(wm.wm.windows, i)
			table.insert(wm.wm.windows, window_local)
			window_local.node:place_top()
		end
	else
		self.obj:clear_focus()
		self.focused = nil
	end
end

-- seat pointer move request
function m.Seat:pointer_move(window_local)
	if self.op == nil then
		self:focus(window_local)
		self.obj:op_start_pointer()
		self.op = {
			type = "move",
			window = window_local,
			start = { x = window_local.x, y = window_local.y },
			dx = 0,
			dy = 0,
		}
	end
end

-- seat pointer resize request
function m.Seat:pointer_resize(window_local, edges)
	if self.op == nil then
		self:focus(window_local)
		window_local.obj:inform_resize_start()
		self.obj:op_start_pointer()
		self.op = {
			type = "resize",
			window = window_local,
			edges = edges,
			start = {
				x = window_local.x,
				y = window_local.y,
				width = window_local.width,
				height = window_local.height,
			},
			dx = 0,
			dy = 0,
		}
	end
end

-- seat action
function m.Seat:action(action)
    local wm = require("wm")
	-- if the action passed == spawn then just use just fork and exec self.arg
	if action == "spawn" then
		if posix.unistd.fork() == 0 then
			if self.arg ~= nil then
				posix.unistd.execp("/bin/sh", { "-c", self.arg })
				os.exit(0)
			end
		end
	elseif action == "close" then
		if self.focused ~= nil then
			self.focused.obj:close()
		end
	elseif action == "focus-next" then
		self:focus(wm.wm.windows[1])
	elseif action == "move" then
		if self.hovered ~= nil then
			self:pointer_move(self.hovered)
		end
	elseif action == "resize" then
		if self.hovered ~= nil then
			self:pointer_resize(self.hovered, { bottom = true, right = true })
        end
	elseif action == "exit" then
		globals.globals["river_window_manager_v1"]:exit_session()
	else
		print("ERROR: m.Seat:action: unimplemented", action)
	end
end

-- seat add pointer binding
function m.Seat:add_pointer_binding(button, mods, action)
	-- From /usr/include/linux/input-event-codes.h
	local button_code = ({ left = 0x110, right = 0x111 })[button]
	local obj = self.obj:get_pointer_binding(button_code, mods)
	local binding = { obj = obj }

	obj:add_listener({
		["pressed"] = function(_)
			self.pending_action = action
		end,
	})
	obj:enable()
	table.insert(self.pointer_bindings, binding)
end

-- seat add xkb_binding
function m.Seat:add_xkb_binding(key, mods, action, arg)
	local keysym = xkbcommon.keysym(key)
	local obj = globals.globals["river_xkb_bindings_v1"]:get_xkb_binding(self.obj, keysym, mods)
	local binding = { obj = obj }

	obj:add_listener({
		["pressed"] = function(_)
			self.pending_action = action
			-- on add keybinding, add another value called arg to the listener
			self.arg = arg
		end,
	})
	obj:enable()
	table.insert(self.xkb_bindings, binding)
end

-- seat manage
function m.Seat:manage()
    local wm = require("wm")
	if self.new then
		self.new = nil
		for _, tbl in ipairs(config.xkb_bindings) do
			self:add_xkb_binding(table.unpack(tbl))
		end

		for _, tbl in ipairs(config.pointer_bindings) do
			self:add_pointer_binding(table.unpack(tbl))
        end

        if not config.config_reload then
            self.layer_shell_seat = globals.globals["river_layer_shell_v1"]:get_seat(self.obj)
            self.layer_shell_seat:add_listener({
                ["focus_none"] = function(_)
                    --rofi closed and dropped "focus-none", so we need to catch it and take away focus from it
                    self.focused = nil
                end,
            })
        end
    end
    -- we do this because when config file is reloaded it tries to call this code again but it cant because
    -- a seat already exists so we check that we arent reloading first

	if self.focused and self.focused.closed then
		self.focused = nil
		if #wm.wm.windows == 0 then
			self:focus(nil)
		end
	end

	if self.interacted then
		self:focus(self.interacted)
		self.interacted = nil
	elseif #wm.wm.windows > 0 then
		local topmost = wm.wm.windows[#wm.wm.windows]
		if self.focused ~= topmost then
			self:focus(topmost)
		end
	end

	if self.pending_action then
		self:action(self.pending_action)
		self.pending_action = nil
	end

	if self.op and self.op.window then
		local op, window_local = self.op, self.op.window

		if window_local.closed then
			self.obj:op_end()
			self.op = nil
		elseif self.op_release then
			if op.type == "resize" then
				window_local.obj:inform_resize_end()
			end
			self.obj:op_end()
			self.op = nil
		elseif op.type == "resize" then
			local width = math.max(
				1,
				op.edges.left and (op.start.width - op.dx)
					or op.edges.right and (op.start.width + op.dx)
					or op.start.width
			)
			local height = math.max(
				1,
				op.edges.top and (op.start.height - op.dy)
					or op.edges.bottom and (op.start.height + op.dy)
					or op.start.height
			)
			window_local.obj:propose_dimensions(width, height)
		end
	end

	self.op_release = nil
end

function m.Seat:render()
	if self.op and self.op.window then
		local op, window_local = self.op, self.op.window
		if self.op.type == "move" then
			window_local:set_position(op.start.x + op.dx, op.start.y + op.dy)
		elseif self.op.type == "resize" then
			local x = op.edges.left and (op.start.x + (op.start.width - window_local.width)) or op.start.x
			local y = op.edges.top and (op.start.y + (op.start.height - window_local.height)) or op.start.y
			window_local:set_position(x, y)
		end
	end
end

function m.Seat:maybe_destroy()
	if self.removed then
		for _, binding in ipairs(self.xkb_bindings) do
			binding.obj:destroy()
		end
		for _, binding in ipairs(self.pointer_bindings) do
			binding.obj:destroy()
		end
		self.obj:destroy()
	else
		return self
	end
end

function m.Seat.listener.removed(self)
	self:get_user_data().removed = true
end
function m.Seat.listener.pointer_enter(self, window_local)
	self:get_user_data().hovered = window_local:get_user_data()
end
function m.Seat.listener.pointer_leave(self)
	self:get_user_data().hovered = nil
end
function m.Seat.listener.window_interaction(self, window_local)
	self:get_user_data().interacted = window_local:get_user_data()
end
function m.Seat.listener.op_delta(self, dx, dy)
	local seat = self:get_user_data()
	seat.op.dx = dx
	seat.op.dy = dy
end
function m.Seat.listener.op_release(self)
	self:get_user_data().op_release = true
end

return m

File: src/seat.c
#include <linux/input-event-codes.h>
#include <stdlib.h>
#include <unistd.h>
#include <wayland-util.h>
#include <xkbcommon/xkbcommon-keysyms.h>
#include <xkbcommon/xkbcommon.h>

#include "actions.h"
#include "config.h"
#include "output.h"
#include "seat.h"
#include "window.h"
#include "wm.h"
#include "xkb.h"

#include <river-layer-shell-v1-client-protocol.h>

void seat_handle_removed(void *data, struct river_seat_v1 *obj) {
    struct Seat *seat = data;
    seat->removed = true;
}

void seat_handle_pointer_enter(void *data, struct river_seat_v1 *obj,
                               struct river_window_v1 *river_window) {
    struct Seat *seat = data;
    seat->hovered = river_window_v1_get_user_data(river_window);
}

void seat_handle_pointer_leave(void *data, struct river_seat_v1 *obj) {
    struct Seat *seat = data;
    seat->hovered = NULL;
}

void seat_handle_window_interaction(void *data, struct river_seat_v1 *obj,
                                    struct river_window_v1 *river_window) {
    struct Seat *seat = data;
    seat->interacted = river_window_v1_get_user_data(river_window);
}

void seat_handle_op_delta(void *data, struct river_seat_v1 *obj, int32_t dx,
                          int32_t dy) {
    struct Seat *seat = data;
    seat->op_dx = dx;
    seat->op_dy = dy;
}

void seat_handle_op_release(void *data, struct river_seat_v1 *obj) {
    struct Seat *seat = data;
    seat->op_release = true;
}

void handle_focused_output_change(struct Seat *seat) {
    fprintf(stdout, "INFO: Changing default layer shell output.\n");
    river_layer_shell_output_v1_set_default(
        seat->focused_output->layer_shell_output);
}

void seat_handle_pointer_position(void *data, struct river_seat_v1 *obj,
                                  int32_t x, int32_t y) {
    struct Seat *seat = data;
    seat->cur_ptr_posx = x;
    seat->cur_ptr_posy = y;

    struct Output *new_output = get_focused_output();
    struct Output *old_output = seat->focused_output;

    if (new_output != old_output) {
        seat->focused_output = new_output;
        if (new_output != NULL) {
            fprintf(stdout, "INFO: Focused output changed.\n");
            handle_focused_output_change(seat);
        }
    }
}

// Ignored events
void seat_handle_wl_seat(void *data, struct river_seat_v1 *obj, uint32_t id) {}
void seat_handle_shell_surface_interaction(
    void *data, struct river_seat_v1 *obj,
    struct river_shell_surface_v1 *river_shell_surface) {}

const struct river_seat_v1_listener river_seat_listener = {
    .removed = seat_handle_removed,
    .wl_seat = seat_handle_wl_seat,
    .pointer_enter = seat_handle_pointer_enter,
    .pointer_leave = seat_handle_pointer_leave,
    .window_interaction = seat_handle_window_interaction,
    .shell_surface_interaction = seat_handle_shell_surface_interaction,
    .op_delta = seat_handle_op_delta,
    .op_release = seat_handle_op_release,
    .pointer_position = seat_handle_pointer_position,
};

void seat_maybe_destroy(struct Seat *seat) {
    if (!seat->removed) {
        return;
    }

    struct XkbBinding *xkb_binding, *xkb_binding_tmp;
    wl_list_for_each_safe(xkb_binding, xkb_binding_tmp, &seat->xkb_bindings,
                          link) {
        xkb_binding_destroy(xkb_binding);
    }

    struct PointerBinding *pointer_binding, *pointer_binding_tmp;
    wl_list_for_each_safe(pointer_binding, pointer_binding_tmp,
                          &seat->pointer_bindings, link) {
        pointer_binding_destroy(pointer_binding);
    }

    river_seat_v1_destroy(seat->obj);
    wl_list_remove(&seat->link);
    free(seat);
}

void seat_focus(struct Seat *seat, struct Window *window) {
    // Focus the top window (if any) when there is no explicit target.
    if (window == NULL && !wl_list_empty(&wm.windows)) {
        window = wl_container_of(wm.windows.prev, window, link);
    }

    if (seat->focused == window) {
        return;
    }

    if (window != NULL) {
        river_seat_v1_focus_window(seat->obj, window->obj);
        river_node_v1_place_top(window->node);
        wl_list_remove(&window->link);
        wl_list_insert(wm.windows.prev, &window->link);
    } else {
        river_seat_v1_clear_focus(seat->obj);
    }

    seat->focused = window;
}

void seat_pointer_move(struct Seat *seat, struct Window *window) {
    seat_focus(seat, window);
    river_seat_v1_op_start_pointer(seat->obj);
    seat->op = SEAT_OP_MOVE;
    seat->op_window = window;
    seat->op_start_x = window->x;
    seat->op_start_y = window->y;
    seat->op_dx = 0;
    seat->op_dy = 0;
}

void seat_pointer_resize(struct Seat *seat, struct Window *window,
                         uint32_t edges) {
    seat_focus(seat, window);
    river_window_v1_inform_resize_start(window->obj);
    river_seat_v1_op_start_pointer(seat->obj);
    seat->op = SEAT_OP_RESIZE;
    seat->op_window = window;
    seat->op_edges = edges;
    seat->op_start_x = window->x;
    seat->op_start_y = window->y;
    seat->op_start_width = window->width;
    seat->op_start_height = window->height;
    seat->op_dx = 0;
    seat->op_dy = 0;
}

void seat_action(struct Seat *seat, enum Action action) {
    switch (action) {
    case ACTION_NONE:
        break;
    case ACTION_SPAWN_SH:
        if (fork() == 0) {
            execlp("/bin/sh", "/bin/sh", "-c", seat->pending_cmd, (char *)0);
        }
        break;
    case ACTION_CLOSE:
        if (seat->focused != NULL) {
            river_window_v1_close(seat->focused->obj);
        }
        break;
    case ACTION_FOCUS_NEXT:
        if (!wl_list_empty(&wm.windows)) {
            // Focus the bottom window
            struct Window *window =
                wl_container_of(wm.windows.next, window, link);
            seat_focus(seat, window);
        }
        break;
    case ACTION_MOVE:
        if (seat->op == SEAT_OP_NONE && seat->hovered != NULL) {
            seat_pointer_move(seat, seat->hovered);
        }
        break;
    case ACTION_RESIZE:
        if (seat->op == SEAT_OP_NONE && seat->hovered != NULL) {
            seat_pointer_resize(seat, seat->hovered,
                                RIVER_WINDOW_V1_EDGES_BOTTOM |
                                    RIVER_WINDOW_V1_EDGES_RIGHT);
        }
        break;
    case ACTION_EXIT:
        river_window_manager_v1_exit_session(window_manager_v1);
        break;
    case ACTION_FULLSCREEN:
        if (seat->focused->fullscreen) {
            fprintf(stdout, "INFO: Leaving fullscreen\n");
            seat_fullscreen_unrequest(seat->focused);
        } else {
            fprintf(stdout, "INFO: Entering fullscreen\n");
            seat_fullscreen_request(seat->focused, get_focused_output());
        }
        break;
    }
}

void fallback_pointerbinds(struct Seat *seat) {
    // Set the fallback pointer binds

    fprintf(stderr, "WARNING: falling back to sane default pointer binds.\n");
    const uint32_t super = RIVER_SEAT_V1_MODIFIERS_MOD4;
    pointer_binding_create(seat, super, BTN_LEFT, ACTION_MOVE);
    pointer_binding_create(seat, super, BTN_RIGHT, ACTION_RESIZE);
}

void fallback_keybinds(struct Seat *seat) {
    // Set the fallback key binds

    fprintf(stderr, "WARNING: falling back to sane default keybinds.\n");
    const uint32_t super = RIVER_SEAT_V1_MODIFIERS_MOD4;
    xkb_binding_create(seat, super, XKB_KEY_Return, ACTION_SPAWN_SH, "foot");
    xkb_binding_create(seat, super, XKB_KEY_d, ACTION_SPAWN_SH,
                       "rofi -show drun");
    xkb_binding_create(seat, super, XKB_KEY_q, ACTION_CLOSE, NULL);
    xkb_binding_create(seat, super, XKB_KEY_n, ACTION_FOCUS_NEXT, NULL);
    xkb_binding_create(seat, super, XKB_KEY_Escape, ACTION_EXIT, NULL);
}

void seat_handle_new(struct Seat *seat) {
    seat->new = false;

    // destroy all keybinds first if they exist
    struct XkbBinding *xkb_binding, *xkb_binding_tmp;
    wl_list_for_each_safe(xkb_binding, xkb_binding_tmp, &seat->xkb_bindings,
                          link) {
        xkb_binding_destroy(xkb_binding);
    }

    struct PointerBinding *pointer_binding, *pointer_binding_tmp;
    wl_list_for_each_safe(pointer_binding, pointer_binding_tmp,
                          &seat->pointer_bindings, link) {
        pointer_binding_destroy(pointer_binding);
    }

    load_config();
    // set pointer bindings
    if (keybind_config.keybinds != NULL) {
        for (size_t i = 0; i < keybind_config.keybinds_len; i++) {
            fprintf(stdout, "INFO: Adding keybind str: %s\n",
                    keybind_config.keybinds[i]);
            parse_and_add_keybind(keybind_config.keybinds[i], seat);
            free(keybind_config.keybinds[i]);
        }
        free(keybind_config.keybinds); // free the array of char*
        keybind_config.keybinds = NULL;
    } else {
        fallback_keybinds(seat);
    }

    if (pointer_config.pointerbinds != NULL) {
        for (size_t i = 0; i < pointer_config.pointerbinds_len; i++) {
            parse_and_add_pointerbind(pointer_config.pointerbinds[i], seat);
            free(pointer_config.pointerbinds[i]);
        }
        free(pointer_config.pointerbinds); // free the array of char*
        pointer_config.pointerbinds = NULL;
    } else {
        fallback_pointerbinds(seat);
    }
}

void seat_manage(struct Seat *seat) {
    if (seat->new) {
        seat_handle_new(seat);
    }

    // If no window was interacted with in the current manage sequence,
    // intentionally pass NULL to ensure the window on top has focus.
    // This is necessary to handle new windows for example.
    seat_focus(seat, seat->interacted);
    seat->interacted = NULL;

    seat_action(seat, seat->pending_action);
    seat->pending_action = ACTION_NONE;

    switch (seat->op) {
    case SEAT_OP_NONE:
        break;
    case SEAT_OP_MOVE:
        if (seat->op_release) {
            river_seat_v1_op_end(seat->obj);
            seat->op = SEAT_OP_NONE;
            seat->op_window = NULL;
            break;
        }
        break;
    case SEAT_OP_RESIZE:
        if (seat->op_release) {
            river_window_v1_inform_resize_end(seat->op_window->obj);
            river_seat_v1_op_end(seat->obj);
            seat->op = SEAT_OP_NONE;
            seat->op_window = NULL;
            break;
        }
        int32_t width = seat->op_start_width;
        int32_t height = seat->op_start_height;
        if ((seat->op_edges & RIVER_WINDOW_V1_EDGES_LEFT) != 0) {
            width -= seat->op_dx;
        }
        if ((seat->op_edges & RIVER_WINDOW_V1_EDGES_RIGHT) != 0) {
            width += seat->op_dx;
        }
        if ((seat->op_edges & RIVER_WINDOW_V1_EDGES_TOP) != 0) {
            height -= seat->op_dy;
        }
        if ((seat->op_edges & RIVER_WINDOW_V1_EDGES_BOTTOM) != 0) {
            height += seat->op_dy;
        }
        river_window_v1_propose_dimensions(seat->op_window->obj,
                                           width > 1 ? width : 1,
                                           height > 1 ? height : 1);
        break;
    }
    seat->op_release = false;
}

void seat_render(struct Seat *seat) {
    // Move and resize stuff, river did this for us.
    switch (seat->op) {
    case SEAT_OP_NONE:
        break;
    case SEAT_OP_MOVE:
        window_set_position(seat->op_window, seat->op_start_x + seat->op_dx,
                            seat->op_start_y + seat->op_dy);
        break;
    case SEAT_OP_RESIZE:;
        int32_t x = seat->op_start_x;
        int32_t y = seat->op_start_y;
        if ((seat->op_edges & RIVER_WINDOW_V1_EDGES_LEFT) != 0) {
            x += seat->op_start_width - seat->op_window->width;
        }
        if ((seat->op_edges & RIVER_WINDOW_V1_EDGES_TOP) != 0) {
            y += seat->op_start_height - seat->op_window->height;
        }
        window_set_position(seat->op_window, x, y);
        break;
    }
}
